### PR TITLE
Fix Javadoc generation on JDK 8

### PR DESCRIPTION
**Pull request by jglick**
# Description

Core was full of Javadoc errors, which JDK 8’s `javadoc` tool rightly complains about. This blocked @kohsuke’s release of 2.54. He worked around it temporarily with 4b5fc6db6fc3f9626497d5bb3396085f22047084 but this is the proper fix.

Please note that what is addressed here are _errors_, which structurally prevent Javadoc from being emitted for various reasons. There are countless _warnings_ like missing parameter documentation which can stay that way forever as far as I am concerned.

The 2.x `plugin-pom` was already set up to ensure that Javadoc is validated during non-release plugin builds, but we never did this before for core.

### Changelog entries

None required.

### Desired reviewers

@reviewbybees & @jenkinsci/code-reviewers

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -44,7 +44,7 @@ for(i = 0; i < buildTypes.size(); i++) {
                             // The -Dmaven.repo.local=${pwd()}/.repository means that Maven will create a
                             // .repository directory at the root of the build (which it gets from the
                             // pwd() Workflow call) and use that for the local Maven repository.
-                            def mvnCmd = "mvn -Pdebug -U clean install ${runTests ? '-Dmaven.test.failure.ignore=true' : '-DskipTests'} -V -B -Dmaven.repo.local=${pwd()}/.repository" 
+                            def mvnCmd = "mvn -Pdebug -U clean install javadoc:javadoc ${runTests ? '-Dmaven.test.failure.ignore=true' : '-DskipTests'} -V -B -Dmaven.repo.local=${pwd()}/.repository"
```
Comment (jglick) [inline Jenkinsfile]: [Done a little differently](https://github.com/jenkinsci/plugin-pom/blob/224bb25f687aa8ffd05092e12d81a10ab46f2f6b/pom.xml#L642-L656) in the plugin POM. At any rate, running Javadoc during CI should prevent regressions from sneaking in which would cause problems for a release.

File: core/src/main/java/hudson/ExtensionListView.java [hunk 2]
```diff
@@ -23,6 +23,7 @@
  */
 package hudson;
 
+import hudson.tasks.UserNameResolver;
```
Comment (jglick) [inline core/src/main/java/hudson/ExtensionListView.java]: Fixes a `@link` at line 49.

File: core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java [hunk 3]
```diff
@@ -44,7 +44,7 @@
  *
  * <p>
  * Have the JavaScript make an AJAX call, to which we respond with 302 redirect. If the reverse proxy
- * is done correctly, this will be handled by {@link #doFoo()}, but otherwise we'll report that as an error.
+ * is done correctly, this will be handled by web methods, but otherwise we'll report that as an error.
```
Comment (jglick) [inline core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java]: Javadoc seems to have fallen out of date with reality.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 4]
```diff
@@ -1053,7 +1053,7 @@ public Calendar due() {
     }
 
     /**
-     * @deprecated Use {@link #getAction(Class)} on {@link AbstractTestResultAction}.
+     * @deprecated Use {@link #getAction(Class)} on {@code AbstractTestResultAction}.
```
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractBuild.java]: Now in `junit-plugin`.

File: core/src/main/java/hudson/model/Build.java [hunk 5]
```diff
@@ -116,7 +116,7 @@ public void run() {
 
     /**
      * @deprecated as of 1.467
-     *      Override the {@link #run()} method by calling {@link #execute(RunExecution)} with
+     *      Override the {@link #run()} method by calling {@link #execute(hudson.model.Run.RunExecution)} with
```
Comment (jglick) [inline core/src/main/java/hudson/model/Build.java]: When referring to nested classes you generally need to use a FQN.

File: core/src/main/java/hudson/model/HealthReport.java [hunk 6]
```diff
@@ -84,7 +84,7 @@
     /**
      * The path to the icon corresponding to this health score or <code>null</code> to use the default icon
      * corresponding to the current health score.
-     * <p/>
+     * <p>
```
Comment (jglick) [inline core/src/main/java/hudson/model/HealthReport.java]: @kohsuke has a lovely blog post somewhere about how evil `<p/>` is…

File: core/src/main/java/hudson/model/PageDecorator.java [hunk 7]
```diff
@@ -48,24 +48,20 @@
  * should implement a new decorator and put {@link Extension} on the class.
  *
  * <h2>Associated Views</h2>
- * <h4>global.jelly</h4>
- * <p>
+ * <h3>global.jelly</h3>
```
Comment (jglick) [inline core/src/main/java/hudson/model/PageDecorator.java]: Out-of-order headers.

File: core/src/main/java/hudson/model/ParameterValue.java [hunk 8]
```diff
@@ -134,10 +134,10 @@ public final String getName() {
      * expected to add more values to this map (or do nothing)
      *
      * <p>
-     * <strike>Environment variables should be by convention all upper case.
```
Comment (jglick) [inline core/src/main/java/hudson/model/ParameterValue.java]: Apparently none of `strike`, `s`, or `del` are allowed.

File: core/src/main/java/hudson/model/Run.java [hunk 9]
```diff
@@ -523,7 +523,7 @@ public boolean isLogUpdated() {
      * and because of that this might not be necessarily in sync with the return value of {@link #isBuilding()} &mdash;
      * an executor holds on to {@link Run} some more time even after the build is finished (for example to
      * perform {@linkplain Run.State#POST_PRODUCTION post-production processing}.)
-     * @see Executables#getExecutor
+     * @see Executor#of
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: I guess?

File: core/src/main/java/hudson/model/listeners/RunListener.java [hunk 10]
```diff
@@ -96,7 +96,7 @@ protected RunListener() {
     public void onCompleted(R r, @Nonnull TaskListener listener) {}
 
     /**
-     * Called after a build is moved to the {@link hudson.model.Run.State#COMPLETED} state.
+     * Called after a build is moved to the {@code Run.State.COMPLETED} state.
```
Comment (jglick) [inline core/src/main/java/hudson/model/listeners/RunListener.java]: `State` is `private` so we cannot link to it.

File: core/src/main/java/jenkins/security/BasicHeaderProcessor.java [hunk 11]
```diff
@@ -38,7 +38,6 @@
  * to authenticate the same header differently and fail.
  *
  * @author Kohsuke Kawaguchi
- * @see ZD-19640
```
Comment (jglick) [inline core/src/main/java/jenkins/security/BasicHeaderProcessor.java]: Was not publicly accessible.

File: pom.xml [hunk 12]
```diff
@@ -354,9 +354,6 @@ THE SOFTWARE.
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-javadoc-plugin</artifactId>
           <version>2.10.3</version>
-          <configuration>
-            <additionalparam>-Xdoclint:none</additionalparam>
-          </configuration>
```
Comment (jglick) [inline pom.xml]: No longer necessary: Javadoc execution should now succeed (with >100 `warning`s but no `error`s).

File: test/pom.xml [hunk 13]
```diff
@@ -245,6 +245,12 @@ THE SOFTWARE.
           </dependency>
         </dependencies>
       </plugin>
+      <plugin>
+        <artifactId>maven-deploy-plugin</artifactId>
```
Comment (oleg-nenashev) [inline test/pom.xml]: unrelated?

**Review by oleg-nenashev — APPROVED** [review]

File: pom.xml [hunk 14]
```diff
@@ -523,7 +520,7 @@ THE SOFTWARE.
         <plugin>
           <groupId>org.jenkins-ci.tools</groupId>
           <artifactId>maven-hpi-plugin</artifactId>
-          <version>1.120</version>
+          <version>1.122-20170411.132219-2</version> <!-- TODO https://github.com/jenkinsci/maven-hpi-plugin/pull/58 -->
```
Comment (jglick) [inline pom.xml]: https://github.com/jenkinsci/maven-hpi-plugin/pull/58 fixes `>` errors in `FormTagLib.java`.

File: test/pom.xml [hunk 15]
```diff
@@ -245,6 +245,12 @@ THE SOFTWARE.
...
+        <configuration>
+          <skip>true</skip>
```
Comment (jglick) [inline test/pom.xml]: Seems we were `deploy`ing an empty JAR on every core release, which is pointless. (The `src/main/java/` was long since moved to `jenkins-test-harness`.)

File: test/pom.xml (hunk 13 above)
Comment (jglick) [inline test/pom.xml]: See above. Related to the extent that Javadoc generation is triggered by running deployment in the release profile. Could be separated if you care.

File: Jenkinsfile (hunk 1 above)
Comment (daniel-beck) [inline Jenkinsfile]: PR builds are building pr-head rather than pr-merge, so we may accidentally merge something that breaks the CI build. So we need to be vigilant about that.

Comment (jglick) [inline Jenkinsfile]: Theoretically. In practice it is unusual for individually passing changes to merge to a breaking change, and in particular it seems unlikely to cause a Javadoc error, which is mostly sensitive to local code (other than changes to members in `@link`). Anyway even if this happened, we would get a clear failure in the `master` job.

File: core/src/main/java/hudson/model/HealthReport.java (hunk 6 above)
Comment (batmat) [inline core/src/main/java/hudson/model/HealthReport.java]: I suspect this comes from the stupid default behaviour IntelliJ has about this :-(

**Review by batmat — APPROVED** [review]
:bee:

File: core/src/main/java/hudson/util/ListBoxModel.java [hunk 16]
```diff
@@ -78,9 +78,9 @@
  * for producing option values. See the following example:
  *
  * <pre>
- * public ListBoxModel doOptionValues(@QueryParameter("value") String value) throws IOException, ServletException {
+ * public ListBoxModel doOptionValues(&#64;QueryParameter("value") String value) throws IOException, ServletException {
```
Comment (MarkEWaite) [inline core/src/main/java/hudson/util/ListBoxModel.java]: I assume that a ```{@code``` wrapper would not work in this case because of the embedded ```@``` character?

**Review by olivergondza — APPROVED** [review]

**Review by MarkEWaite — APPROVED** [review]

File: core/src/main/java/hudson/util/ListBoxModel.java (hunk 16 above)
Comment (jglick) [inline core/src/main/java/hudson/util/ListBoxModel.java]: No, the problem with `@code` is that it does not support nested `{…}` sequences, making it useless for most multiline Java code samples.

**Comment by jglick** [comment]
CI failure looks like a timeout, perhaps.
